### PR TITLE
Configure postgres for production

### LINKED ISSUE #361: Postgresql tuning
This was mentioned somewhere else, but we don't have an issue for it.

Postgresql sets incredibly conservative configuration values by default that are not well-suited to production servers.

There are quite a few "recommended" tuning settings but most of them are based on the resources of the system being tuned (available RAM, etc). I've anecdotally seen references to ~30% performance increases just by applying recommended settings.

Given that we seem to be hitting some performance issues, maybe we should prioritise a spike on this at some point for some quick wins.

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
Closes #361 

This is a draft playbook to configure PostgreSQL for production servers by setting some recommended values.

It might be a good idea to get the Datadog paid plan first, as discussed here: https://community.openfoodnetwork.org/t/making-operations-a-first-class-citizen/1601/5

Then we could try provisioning this configuration to one production server and get some data on the results.

### CODE REVIEW

File: playbooks/postgres_tuning.yml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+    - name: Add tuning configuration
+      copy:
+        content: |
+          shared_buffers = {{ ( ansible_memtotal_mb / 4 ) | int }}MB
+          effective_cache_size = {{ ( ( ansible_memtotal_mb / 4 ) * 3 ) | int }}MB
+          maintenance_work_mem = {{ ( ansible_memtotal_mb / 16 ) | int }}MB
+          wal_buffers = 16MB
+          min_wal_size = 1GB
+          max_wal_size = 2GB
```
Comment (sauloperez) [inline playbooks/postgres_tuning.yml]: Can we make at least these two relative to `ansible_memtotal_mb` as well? We're working on getting a production server with the standard specs at Katuma but we're not there yet. 

I checked pgtune and still with a 2 GB RAM it sets max_wal_size = 2GB :man_shrugging:

**Comment by sauloperez** [comment]
I also think we will need to hold this until we reach consensus on https://github.com/openfoodfoundation/ofn-install/issues/367

**Comment by Matt-Yorkley** [comment]
> What about using an Ansible role such as https://github.com/geerlingguy/ansible-role-postgresql and just configure things through per-host variables if needs be? IMO it's simpler that way compared to having multiple conf files deviating from the standard.

I think using that postgres galaxy role you mentioned would be a big improvement, and there are several issues around that could be grouped under a general postgres refactor: #99, #72, #227.

On the other hand, that role won't optimise the configuration at all, which still I think we should try to do...

**Comment by Matt-Yorkley** [comment]
I put this up as a general proposal for discussion more than anything. Maybe we could remove the `wal_buffers_min` and `wal_buffers_max`? 

UK would definitely like to try some optimisation from minimal to recommended settings, so what I'm trying to avoid with this PR is:
- manually adding a non-standardised configuration to UK production that doesn't appear in ofn-install (and forgetting it's there)
- having to re-add a non-standardised configuration if we migrate the server
- having to manually re-configure hard coded values higher if we add more ram

**Comment by sauloperez** [comment]
> On the other hand, that role won't optimise the configuration at all, which still I think we should try to do...

Of course not but it would be a great foundation to make it easy to manage and iterate its configuration. If it is affordable I would that first we start from nothing anyway.

**Comment by sauloperez** [comment]
Would you guys mind waiting to have a refactored provision.yml using https://github.com/geerlingguy/ansible-role-postgresql? It does it better than us and I feel it'll make a lot easier to iterate on postgres config.

**Comment by Matt-Yorkley** [comment]
> Would you guys mind waiting to have a refactored provision.yml using https://github.com/geerlingguy/ansible-role-postgresql? It does it better than us and I feel it'll make a lot easier to iterate on postgres config. It was my next step.

Sounds good, and we can define a postgres version at the same time. When I was playing with ofn-install in travis it installed PG 11 by default as we don't specify the version _anywhere_.

Okay, I'll do it this week. We can wait until then to merge this. :+1:

**Comment by Matt-Yorkley** [comment]
I've updated this to be a bit cleaner, and it now depends on #400.

I think use of this `conf.d` pattern for storing custom settings is the cleanest way to manage configurations.

Screenshot-2019-04-16-21-56:48

**Comment by Matt-Yorkley** [comment]
> What do these values mean and why are they better than the default?

Similar to the `unicorn_workers` experiments, the default postgres settings are generally regarded to be a bit low for production servers. The idea here is that we can apply some recommended settings that should enhance performance, and monitor the results with Datadog.

Mostly I'm trying to look for ways we can get better performance from what we have, and avoid the downtime we're starting to see regularly on French and UK staging.

Some further reading here: 
https://www.revsys.com/writings/postgresql-performance.html
https://leopard.in.ua/2014/03/24/pgtune-for-postgresql

The playbook is optional, so we could run it on one server as a test and see the results. The playbook also includes a task for removing the configuration from a server.

**Comment by sauloperez** [comment]
I think the main idea @mkllnk is to increase the PostgreSQL's mem usage so it doesn't hit disk that often. This will improve performance in all sorts of tasks the server performs. And it turns out our servers have a considerable amount of RAM available.

File: playbooks/postgres_tuning.yml (hunk 1 above)
Comment (Matt-Yorkley) [reply playbooks/postgres_tuning.yml]: `min_wal_size` and `max_wal_size` have been removed, so they will stay at their defaults. :+1:

**Pull request merged**